apps/email/js/cards/cmp/autocomplete.js

```diff
@@ -0,0 +1,565 @@
+      span.textContent = matchString;
+      node.appendChild(span);
+    }
+
+    var endString = text.substring(endIndex);
+    if (endString) {
+      node.appendChild(document.createTextNode(endString));
+    }
+  } else {
+    node.textContent = text;
+  }
+}
+
+/**
+ * Given a node, if it or an ancestor has the CSS className, return the node
+ * with that className. Returns null if not a match.
+ * @param  {Node} node
+ * @return {Node}
+ */
+function findAncestorWithClass(node, className) {
+  if (node.classList && node.classList.contains(className)) {
+    return node;
+  }
+
+  while ((node = node.parentNode)) {
+    if (node.classList && node.classList.contains(className)) {
+      return node;
+    }
+  }
+
+  return null;
+}
+
+/**
+ * Given a node, if it or an ancestor is an autocomplete item in the
+ * autocomplete list, return the autcomplete item node. Returns null if not a
+ * match.
+ * @param  {Node} node
+ * @return {Node}
+ */
+function getAutocompleteItem(node) {
+  return findAncestorWithClass(node, 'cmp-autocomplete-item');
+}
+
+/**
+ * Autocomplete custom element that contains a list of elements that show the
+ * autocomplete matches.
+ *
+ * This custom element relies on a CSS class-based API to bind to nodes outside
+ * of this autocomplete. The CSS/HTML structure expected by this autocomplete:
+ *
+ * - An element with the CSS class of cmp-autocomplete-origin" that is the the
+ * master scroll area for the view. It should be an ancestor node to the
+ * autocomplete and also contain the 'cmp-autocomplete-input-list' element that
+ * has all the input elements. This element is used as the "origin" for the
+ * style.top placement of the autocomplete list.
+ *
+ * - An element that holds all the input fields, with the CSS class of
+ * "cmp-autocomplete-input-list". That inputArea will be translateY'd to make
+ * sure the current input focus is visible at the top of the view.
+ *
+ * The autocomplete custom element also expects a "data-source" HTML attribute
+ * to be set on it, which indicates the module ID to use as the data source for
+ * the matches.
+ *
+ * The owner of this custom element instance (typically a view card), should
+ * override the getExistingEntries() method on the instance to return the
+ * current entries, to avoid dupes showing up in subsequent autocomplete
+ * matches.
+ */
+return [
+  require('../base')(),
+  {
+    createdCallback: function() {
+      //TODO: master hack only. In email-model-instance branch, this is not
+      //needed. In the branch, there is a separate 'base' that elements can
+      //mix in that is not assumed to be a card. For now, this is an equivalent
+      //result.
+      this.classList.remove('card');
+
+      this.lastTranslateY = 0;
+      this.inputAreaTranslated = Promise.resolve();
+
+      this.runQuery = this.runQuery.bind(this);
+      this.onFocus = this.onFocus.bind(this);
+      this.onDocumentEvent = this.onDocumentEvent.bind(this);
```

Comment:
General thought not for this specific patch: I wonder if it would be reasonable to just auto-bind() all of the methods on our custom elements.  That's what React.createElement does (the old, pre-"hey just use ES2015 classes" idiom, which I still use), and it's handy since it avoids the boilerplate and the foot-gun of failing to do this.